File: Solutions/PE002_even_fibonacci_numbers/even_fibonacci_numbers.py
#!/bin/python3

# #https://www.hackerrank.com/contests/projecteuler/challenges/euler002

from functools import lru_cache


@lru_cache(maxsize=None)
def fib(n):
    """
    returns the nth fibonacci number
    """
    if n == 1:
        return 1
    if n == 2:
        return 2
    if n > 2:
        return fib(n - 1) + fib(n - 2)


def even_fibonacci_numbers(n):
    """
    returns the sum of the even valued fibonacci numbers less than or equal to n
    """
    even_fib_less_than_n = []
    i = 1
    fib_i = 1
    while fib_i <= n:
        if fib_i % 2 == 0:
            even_fib_less_than_n.append(fib_i)
        i += 1
        fib_i = fib(i)
    return sum(even_fib_less_than_n)


if __name__ == '__main__':
    print(even_fibonacci_numbers(4000000))
    # t = int(input().strip())
    # for a0 in range(t):
    #     n = int(input().strip())
    #     even_fib_lessthan_n = []
    #     i = 1
    #     fib_i = 1
    #     while fib_i <= n:
    #         if fib_i % 2 == 0:
    #             even_fib_lessthan_n.append(fib_i)
    #         i += 1
    #         fib_i = fib(i)
    #     print(sum(even_fib_lessthan_n))

File: Solutions/PE002_even_fibonacci_numbers/cpp/EvenFibonacciNumbers.cpp
#include <iostream>
#include <vector>
#include <map>

//https://www.hackerrank.com/contests/projecteuler/challenges/euler002

using namespace std;

map<long, long> cache;

long int Fib(long int n) {
	if (cache.find(n) != cache.end())
		return cache[n];
	else {
		if (n == 1)
			return 1;
		else if (n == 2)
			return 2;
		else if (n > 2) {
			cache[n] = Fib(n - 1) + Fib(n - 2);
			return cache[n];
		}
		else
			return 1;
	}
}

long int EvenFib(long int n) {
	//returns the sum of the even valued fibonacci numbers less than or equal to n
	vector<long int> EvenFibLessThann;
	long int i = 1;
	long int fib_i = 1;
	while (fib_i <= n) {
		if (fib_i % 2 == 0)
			EvenFibLessThann.push_back(fib_i);
		i++;
		fib_i = Fib(i);
	}
	long int sum_of_elems = 0;
	for (auto i : EvenFibLessThann)
		sum_of_elems += i;
	return sum_of_elems;
}

int main() {
	int t;
	cin >> t;
	for (int a0 = 0; a0 < t; a0++) {
		long int n;
		cin >> n;
		cout << EvenFib(n) << endl;
	}
}
